WINTERMUTE: Add keymappings

lolbot-iichan opened the pull request.

bgK commented on engines/wintermute/detection.cpp:
```diff
@@ -210,6 +214,214 @@ class WintermuteMetaEngine : public AdvancedMetaEngine {
+		
+		Common::KeymapArray result = Keymap::arrayOf(engineKeyMap);
+
+		Common::String targetString = target;
```
The target is user-controlled. It's possible to edit it when adding a game. Is fetching the gameid from the configuration manager not working for you? `Common::String gameId = ConfMan.get("gameid", target);`

lolbot-iichan replied:
Thanks! Never noticed "gameid" property before.

lolbot-iichan commented:
Current state of this PR:
1. Standard in-game controls & game-specific controls for all 90+ games listed at https://wiki.scummvm.org/index.php?title=Wintermute/Controls are all added to detection.cpp, except for "onehelluvaday" demo, which features an action binded to Shift keypress. It seems that currently it's impossible to either add default input mapping for Shift, or assign it manually with GUI. Original inputs are all marked with "//original [keyboard|mouse|joy]" comments.

2. Text input engine-hardcoded controls are still hardcoded, I don't plan to change it in this PR. 

3. Extra keyboard input mappings (UP+DOWN or PGUP+PGDN) are added for a couple of games. They are marked with "// extra keyboard". Ideas begind assigning extra keyboard keys:
* original keymaps are never changed, if both UP+DOWN and PGUP+PGDN is already used, then nothing is added.
* if game is using mouse wheel and does not use UP+DOWN or PGUP+PGDN pair -> map one of them to the mouse wheel

4. Extra joystick mappings are added for all games. They are marked with "// extra joy". Ideas behind assigning JOY buttons:
* original keymaps are never changed
* All games are using LMB=JOY_A, RMB=JOY_B, ESC=JOY_X.
* Game's most useful action (usually, HINTS or INV) is binded to JOY_Y.
* If game has paired controls like scrolling, changing pages, walking, -> assign those to JOY arrows.
* A few games features unpaired actions binded to JOY_LEFT and/or JOY_RIGHT (like, Face Noir's INV on JOY_LEFT and HINTS on JOY_Y). I'm not sure if it's ok or not to call commands with DPAD keys. Maybe JOY_R+Y would be nice to have, maybe not - not use...
* Enter is curretly unmapped to JOY. It would be really nice if JOY_R+JOY_X would be allowed for Enter. Almost all of listed games are using Enter at windows like "Are you sure you want %s", but I'm not sure if it's a shortcut for OK button in all the games or some games are not fully playable without ability to press Enter. I'd better have a JOY button for it.
* All cheats are unmapped to JOY. It would be really nice if JOY_R+JOY_B combo would be allowed for cheats & secrets. FoxTail has 2 cheats, for JOY_R+JOY_A would be useful as well.

5. Extra mouse input mappings (MOUSE_WHEELUP+MOUSE_WHEELDOWN,  MOUSE_MIDDLE) are added. They are marked with "// extra mouse".  Ideas behind assigning extra MOUSE buttons:
* original keymaps are never changed, if MOUSE_MIDDLE and/or MOUSE_WHEELUP+MOUSE_WHEELDOWN are already used, rules below does not apply.
* If MOUSE_RIGHT is not actually used, it may be remapped (used only for "Wilma Tetris", which does not use RightClick event).
* JOY_Y has some action -> assign MOUSE_ MIDDLE to the same action (except for Face Noir, where INV action is mapped to MOUSE_MIDDLE in original keymap, while I chose HINTS to be assigned to JOY_Y).
* If game has paired controls like scrolling, changing pages, switching between 2 GUI types -> assign those action assigned to MOUSE_WHEELUP/MOUSE_WHEELDOWN arrows

bgK commented:
Humongous work! We can merge this whenever you feel it's ready.

> It seems that currently it's impossible to either add default input mapping for Shift, or assign it manually with GUI. 

I'll look into fixing that, there is no reason not to be able to remap the modifier keys.

lolbot-iichan commented:
>Humongous work! We can merge this whenever you feel it's ready.

Thanks! I guess I'll wait for #2052 to complete and will do a rebase, since my branch conflicts with that PR.

Meanwhile, current open questions:
* Should I add JOY_R+JOY_X for Enter or leave it unmapped?
* Should I add JOY_R+JOY_B for secrets or leave them unmapped?
If at least one of those is yes, then what is the syntax? `act->addDefaultInputMapping("JOY_R+JOY_X")` does not seem to work for me (GUI does not list this combo).

bgK commented:
> Should I add JOY_R+JOY_X for Enter or leave it unmapped?
> Should I add JOY_R+JOY_B for secrets or leave them unmapped?

At the moment, joystick button combos are not expressible using the keymapper. I suggest leaving those actions unmapped with a comment describing the ideal bindings. Users can map them to device specific buttons (triggers, stick press, right stick axes) if needed.

bgK commented:
> > It seems that currently it's impossible to either add default input mapping for Shift, or assign it manually with GUI.
> 
> I'll look into fixing that, there is no reason not to be able to remap the modifier keys.

Done in 29dd15af0c3ddb8f2b130c60b35ac0d9e115d695.

bluegr commented on engines/wintermute/detection.cpp:
```diff
@@ -210,6 +214,1425 @@ class WintermuteMetaEngine : public AdvancedMetaEngine {
+		act = new Action("RETURN", _("Confirm"));
+		act->setKeyEvent(KeyState(KEYCODE_RETURN, ASCII_RETURN));
+		act->addDefaultInputMapping("RETURN"); // original keyboard
+//TODO: extra joy control, e.g. "JOY_R+JOY_X"
```
Please, indent all of these TODO comments properly (more follow below, no reason to repeat this comment)

lolbot-iichan replied:
Intended in line with code with tabs.

bluegr commented:
Extraordinary work! Well done!

Could you please fix the indentation of the TODO comments? Other than that, it's an amazing effort, kudos!

lolbot-iichan commented:
Rebased, removed debugger handling (it's already done at #2052), added Shift hotkey, moved TODOs to the right. 

> Done in [29dd15a](https://github.com/scummvm/scummvm/commit/29dd15af0c3ddb8f2b130c60b35ac0d9e115d695).

Thanks, supported that at 9e267aa.

>Could you please fix the indentation of the TODO comments?

Fixed at 9e267aa.

>We can merge this whenever you feel it's ready.

I feel it's ready.

ccawley2011 commented on engines/wintermute/detection.cpp:
```diff
@@ -210,6 +214,1421 @@ class WintermuteMetaEngine : public AdvancedMetaEngine {
+			act->setRightClickEvent();
+			act->addDefaultInputMapping("MOUSE_RIGHT"); // original mouse (globally)
+			act->addDefaultInputMapping("LEFTBRACKET"); // original keyboard (ingame)
+			act->addDefaultInputMapping("MOUSE_WHEEL_UP"); // original mouse
```
This doesn't seem to match what is described on the [Wiki article](https://wiki.scummvm.org/index.php?title=Wintermute/Controls#Pizza_Morgana:_Episode_1_-_Monsters_and_Manipulations_in_the_Magical_Forest), which lists "[" for selecting the previous action and "]" for selecting the next action.

lolbot-iichan replied:
Thanks for pointing this. Took a deep look at Pizza Morgana and ZBang code.
Wiki version was correct, both those games share code that handles MouseWheelUp, RightClick, RightDoubleClick and "]" as "Select Next Action", and MouseWheelDown & "[" as "Select Previous Action". Fixed.

ccawley2011 commented on engines/wintermute/detection.cpp:
```diff
@@ -210,6 +214,1421 @@ class WintermuteMetaEngine : public AdvancedMetaEngine {
+			gameKeyMap->addAction(act);
+
+			act = new Action("ACTPRV", _("Previous action"));
+			act->setRightClickEvent();
```
In addition to what I said about Pizza Morgana, both "ACTNXT" and "ACTPRV" are set to send right click events, which doesn't seem correct.

lolbot-iichan replied:
Thanks. Fixed, see above.

lolbot-iichan commented on engines/wintermute/detection.cpp:
```diff
@@ -210,6 +214,1425 @@ class WintermuteMetaEngine : public AdvancedMetaEngine {
+			gameId == "rhiannon" ||
+			gameId == "shinestar"
+		) {
+			act = new Action("SKIP", _("Skip"));
```
Thanks, fixed! Maybe we should also introduce more Standard Actions for things like "HINT", "WTF", "ACTNXT", "PAGEDN", "SCRLUP", etc?

bgK replied:
Yes, feel free to add some. I'm not convinced standard actions are very useful at the moment. We set the default bindings for mouse+keyboard and gamepad when we define the actions. The backend changing the defaults for the standard actions is very likely to result in nonsensical / conflicting keymaps.

One possible future change is the evolution of the virtual keyboard into some sort of more general touch oriented on-screen input method. Having standard actions in that context would be nice as we could bundle a set of corresponding icons.

lolbot-iichan commented on engines/wintermute/detection.cpp:
```diff
@@ -210,6 +214,1425 @@ class WintermuteMetaEngine : public AdvancedMetaEngine {
+			act->addDefaultInputMapping("JOY_Y"); // extra joy
+			gameKeyMap->addAction(act);
+
+			act = new Action("HINT", _("Show inventory"));
```
Thanks for spotting this, fixed!

lolbot-iichan commented on engines/wintermute/detection.cpp:
```diff
@@ -210,6 +214,1425 @@ class WintermuteMetaEngine : public AdvancedMetaEngine {
+//TODO: extra joy control, e.g. "JOY_R+JOY_B"
+			gameKeyMap->addAction(act);
+
+			act = new Action("DBG", _("Debug print"));
```
Renamed to "DBGTXT", thanks for pointing this.
Should we introduce some conventions to avoid things like this in the future? Like, "non-standard actions IDs should always be 6+ characters long"?

bgK replied:
Well the action ids are namespaced by the keymap names. There was no real conflict here. The only issue is if a backend remaps a standard action that happens to have the same name as game action. I'm not sure it's worth the effort of enforcing a naming convention.

lolbot-iichan commented:
Fixed errors from above + splited "5ma||5ld||dirtysplit||projectjoe" keymap into two, since they slightly different original controls + fixed action ID collision at Alpha Polaris.

bluegr commented:
Great work, once again! Thanks for fixing the pending issues. Merging
